File: bot-intelligence/src/main/java/gg/mineral/bot/ai/goal/DrinkPotionGoal.java
package gg.mineral.bot.ai.goal;

import gg.mineral.bot.api.controls.Key;
import gg.mineral.bot.api.controls.MouseButton;
import gg.mineral.bot.api.entity.living.player.ClientPlayer;

import gg.mineral.bot.api.event.Event;
import gg.mineral.bot.api.event.peripherals.MouseButtonEvent;
import gg.mineral.bot.api.goal.Goal;

import gg.mineral.bot.api.instance.ClientInstance;
import gg.mineral.bot.api.inv.item.Item;

import gg.mineral.bot.api.screen.type.ContainerScreen;
import lombok.Getter;
import lombok.val;

@Getter
public class DrinkPotionGoal extends Goal {

    private boolean inventoryOpen = false, drinking = false;

    @Override
    public boolean shouldExecute() {
        boolean shouldExecute = drinking || canSeeEnemy() && hasDrinkablePotion();
        info(this, "Checking shouldExecute: " + shouldExecute);
        return shouldExecute;
    }

    public DrinkPotionGoal(ClientInstance clientInstance) {
        super(clientInstance);
        info(this, "DrinkPotionGoal initialized");
    }

    private boolean hasDrinkablePotion() {
        val fakePlayer = clientInstance.getFakePlayer();
        val inventory = fakePlayer.getInventory();

        if (inventory == null) {
            warn(this, "Inventory is null");
            return false;
        }

        boolean hasDrinkablePotion = inventory.containsPotion(potion -> !potion.isSplash());
        info(this, "Has drinkable potion: " + hasDrinkablePotion);
        return hasDrinkablePotion;
    }

    private boolean canSeeEnemy() {
        val fakePlayer = clientInstance.getFakePlayer();
        val world = fakePlayer.getWorld();

        if (world == null) {
            warn(this, "World is null");
            return false;
        }

        boolean canSeeEnemy = world.getEntities().stream()
                .anyMatch(entity -> entity instanceof ClientPlayer
                        && !clientInstance.getConfiguration().getFriendlyUUIDs().contains(entity.getUuid()));
        info(this, "Checking canSeeEnemy: " + canSeeEnemy);
        return canSeeEnemy;
    }

    private void drinkPotion() {
        drinking = true;
        success(this, "Started drinking potion");
    }

    private void switchToDrinkablePotion() {
        info(this, "Switching to a drinkable potion");
        var potionSlot = -1;
        val fakePlayer = clientInstance.getFakePlayer();
        val inventory = fakePlayer.getInventory();

        if (inventory == null) {
            warn(this, "Inventory is null");
            return;
        }

        for (int i = 0; i < 36; i++) {
            val itemStack = inventory.getItemStackAt(i);
            if (itemStack == null)
                continue;
            val item = itemStack.getItem();
            if (item.getId() == Item.POTION) {
                val potion = itemStack.getPotion();
                if (potion.isSplash())
                    continue;

                potionSlot = i;
                break;
            }
        }

        if (potionSlot > 8) {
            if (!inventoryOpen) {
                inventoryOpen = true;
                pressKey(10, Key.Type.KEY_E);
                info(this, "Opened inventory");
                return;
            }

            val screen = clientInstance.getCurrentScreen();

            if (screen == null) {
                inventoryOpen = false;
                warn(this, "Screen is null; closing inventory");
                return;
            }

            val inventoryContainer = fakePlayer.getInventoryContainer();

            if (inventoryContainer == null) {
                inventoryOpen = false;
                warn(this, "Inventory container is null; closing inventory");
                return;
            }

            val slot = inventoryContainer.getSlot(inventory, potionSlot);

            if (slot == null) {
                inventoryOpen = false;
                pressKey(10, Key.Type.KEY_ESCAPE);
                warn(this, "Potion slot is null; closing inventory");
                return;
            }

            if (screen instanceof ContainerScreen containerScreen) {
                val x = containerScreen.getSlotX(slot);
                val y = containerScreen.getSlotY(slot);

                val currX = getMouseX();
                val currY = getMouseY();

                if (currX != x || currY != y) {
                    setMouseX(x);
                    setMouseY(y);
                    info(this, "Moving mouse to potion slot at x: " + x + ", y: " + y);
                } else {
                    pressKey(10, Key.Type.KEY_8);
                    success(this, "Moved to end slot");
                }
            }
            return;
        }

        if (inventoryOpen) {
            inventoryOpen = false;
            pressKey(10, Key.Type.KEY_ESCAPE);
            info(this, "Closing inventory after switching potion");
            return;
        }

        pressKey(10, Key.Type.valueOf("KEY_" + (potionSlot + 1)));
        success(this, "Switched to potion slot: " + (potionSlot + 1));
    }

    @Override
    public void onTick() {
        val fakePlayer = clientInstance.getFakePlayer();
        val inventory = fakePlayer.getInventory();

        if (inventory == null) {
            warn(this, "Inventory is null on tick");
            return;
        }

        val itemStack = inventory.getHeldItemStack();

        if (drinking && (itemStack == null || itemStack.getItem().getId() != Item.POTION)) {
            drinking = false;
            info(this, "Stopped drinking as no potion is held");
        }

        val rmbHeld = getButton(MouseButton.Type.RIGHT_CLICK).isPressed();

        if (drinking && !rmbHeld) {
            pressButton(MouseButton.Type.RIGHT_CLICK);
            info(this, "Pressed RIGHT_CLICK for drinking");
        }

        if (!drinking && rmbHeld) {
            unpressButton(MouseButton.Type.RIGHT_CLICK);
            info(this, "Unpressed RIGHT_CLICK as drinking stopped");
        }

        if (drinking || !delayedTasks.isEmpty()) {
            return;
        }

        if (itemStack != null && itemStack.getItem().getId() == Item.POTION) {
            schedule(() -> drinkPotion(), 100);
            info(this, "Scheduled drinkPotion task");
        } else {
            schedule(() -> switchToDrinkablePotion(), 100);
            info(this, "Scheduled switchToDrinkablePotion task");
        }
    }

    @Override
    public boolean onEvent(Event event) {
        if (event instanceof MouseButtonEvent mouseButtonEvent) {
            if (drinking && mouseButtonEvent.getType() == MouseButton.Type.RIGHT_CLICK
                    && !mouseButtonEvent.isPressed()) {
                info(this, "Ignoring RIGHT_CLICK release event while drinking");
                return true;
            }
        }
        return false;
    }

    @Override
    public void onGameLoop() {
    }
}

File: bot-intelligence/src/main/kotlin/gg/mineral/bot/ai/goal/DrinkPotionGoal.kt
package gg.mineral.bot.ai.goal

import gg.mineral.bot.ai.goal.type.InventoryGoal
import gg.mineral.bot.api.controls.Key
import gg.mineral.bot.api.controls.MouseButton
import gg.mineral.bot.api.entity.living.ClientLivingEntity
import gg.mineral.bot.api.entity.living.player.ClientPlayer
import gg.mineral.bot.api.event.Event
import gg.mineral.bot.api.event.peripherals.MouseButtonEvent
import gg.mineral.bot.api.goal.Sporadic
import gg.mineral.bot.api.goal.Timebound
import gg.mineral.bot.api.instance.ClientInstance
import gg.mineral.bot.api.inv.item.Item
import gg.mineral.bot.api.inv.item.ItemStack
import gg.mineral.bot.api.inv.potion.Potion
import gg.mineral.bot.api.screen.type.ContainerScreen
import gg.mineral.bot.api.util.fastArcTan

class DrinkPotionGoal(clientInstance: ClientInstance) : InventoryGoal(clientInstance), Sporadic, Timebound {
    override var executing: Boolean = false
    override var startTime: Long = 0
    override val maxDuration: Long = 100
    private var drinking = false

    override fun shouldExecute(): Boolean {
        val shouldExecute = canSeeEnemy() && hasDrinkablePotion()
        logger.debug("Checking shouldExecute: $shouldExecute")
        return shouldExecute
    }

    override fun onStart() {
        pressKey(Key.Type.KEY_W, Key.Type.KEY_LCONTROL)
        unpressKey(Key.Type.KEY_S, Key.Type.KEY_A, Key.Type.KEY_D)
    }

    init {
        logger.debug("DrinkPotionGoal initialized")
    }

    private fun hasDrinkablePotion(): Boolean {
        val fakePlayer = clientInstance.fakePlayer
        val inventory = fakePlayer.inventory

        val hasDrinkablePotion = inventory.containsPotion {
            isValidPotion(it)
        }
        logger.debug("Has drinkable potion: $hasDrinkablePotion")
        return hasDrinkablePotion
    }

    private fun canSeeEnemy(): Boolean {
        val fakePlayer = clientInstance.fakePlayer
        val world = fakePlayer.world

        val canSeeEnemy = world.entities
            .any {
                it is ClientPlayer
                        && !clientInstance.configuration.friendlyUUIDs.contains(it.uuid)
            }
        logger.debug("Checking canSeeEnemy: $canSeeEnemy")
        return canSeeEnemy
    }

    private fun angleAwayFromEnemies(): Float {
        val fakePlayer = clientInstance.fakePlayer
        val world = fakePlayer.world

        val enemy = world.entities
            .minByOrNull {
                if (it is ClientLivingEntity && !clientInstance.configuration.friendlyUUIDs.contains(it.uuid))
                    it.distance3DTo(fakePlayer)
                else Double.MAX_VALUE
            } ?: return fakePlayer.yaw
        val x: Double = enemy.x - fakePlayer.x
        val z: Double = enemy.z - fakePlayer.z

        var yaw = Math.toDegrees(-fastArcTan(x / z)).toFloat()
        if (z < 0.0 && x < 0.0) yaw = (90.0 + Math.toDegrees(fastArcTan(z / x))).toFloat()
        else if (z < 0.0 && x > 0.0) yaw = (-90.0 + Math.toDegrees(fastArcTan(z / x))).toFloat()
        return yaw + 180.0f
    }

    private fun distanceAwayFromEnemies(): Double {
        val fakePlayer = clientInstance.fakePlayer
        val world = fakePlayer.world

        return world.entities
            .minOfOrNull {
                if (it is ClientLivingEntity && !clientInstance.configuration.friendlyUUIDs.contains(it.uuid))
                    it.distance3DTo(fakePlayer)
                else Double.MAX_VALUE
            } ?: Double.MAX_VALUE
    }

    private fun isValidPotion(potion: Potion): Boolean {
        val fakePlayer = clientInstance.fakePlayer
        // TODO: exclude negative potions
        for (effect in potion.effects) if (fakePlayer.isPotionActive(effect.potionID)) return false
        return !potion.isSplash && potion.effects.isNotEmpty()
    }

    private fun isValidPotion(itemStack: ItemStack): Boolean {
        if (itemStack.item.id != Item.POTION) return false
        val potion = itemStack.potion ?: return false
        return isValidPotion(potion)
    }

    private fun getPotionSlot(): Int {
        var potionSlot = -1
        val fakePlayer = clientInstance.fakePlayer
        val inventory = fakePlayer.inventory

        // Look for a non-splash potion in one of the 36 slots
        invLoop@ for (i in 0..35) {
            val itemStack = inventory.getItemStackAt(i) ?: continue
            if (isValidPotion(itemStack)) {
                potionSlot = i
                break
            }
        }
        return potionSlot
    }

    override fun onTick(tick: Tick) {
        val potionSlot = getPotionSlot()
        val fakePlayer = clientInstance.fakePlayer
        val inventory = fakePlayer.inventory

        tick.finishIf("No Valid Potion Found", potionSlot == -1)

        tick.prerequisite("In Hotbar", potionSlot <= 8) {
            moveItemToHotbar(potionSlot, inventory)
        }

        tick.prerequisite(
            "Inventory Closed",
            clientInstance.currentScreen !is ContainerScreen
        ) { pressKey(10, Key.Type.KEY_ESCAPE) }

        tick.prerequisite("Correct Hotbar Slot Selected", inventory.heldSlot == potionSlot) {
            pressKey(10, Key.Type.valueOf("KEY_" + (potionSlot + 1)))
        }

        tick.finishIf("Not Holding Valid Potion", inventory.heldItemStack?.let { isValidPotion(it) } == false)

        tick.finishIf("Potion Not Needed", !shouldExecute())

        tick.prerequisite("Drinking", drinking && getButton(MouseButton.Type.RIGHT_CLICK).isPressed) {
            pressButton(MouseButton.Type.RIGHT_CLICK)
            drinking = true
        }

        tick.execute {
            if (distanceAwayFromEnemies() < 16) {
                setMouseYaw(angleAwayFromEnemies())
                pressKey(Key.Type.KEY_SPACE, Key.Type.KEY_W, Key.Type.KEY_LCONTROL)
            }
        }
    }

    override fun onEnd() {
        drinking = false
        unpressButton(MouseButton.Type.RIGHT_CLICK)
        unpressKey(Key.Type.KEY_SPACE)
    }

    override fun onEvent(event: Event): Boolean {
        if (event is MouseButtonEvent) {
            if (drinking && event.type == MouseButton.Type.RIGHT_CLICK && !event.pressed) {
                logger.debug("Ignoring RIGHT_CLICK release event while drinking")
                return true
            }
        }
        return false
    }

    public override fun onGameLoop() {}
}
